iothub_client/samples/pnp/common/pnp_dps.c

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef USE_PROV_MODULE
+#error "Missing cmake flag for DPS"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "iothub_device_client.h"
+#include "iothubtransportmqtt.h"
+#include "pnp_device_client_helpers.h"
+
+#include "azure_c_shared_utility/threadapi.h"
+#include "azure_c_shared_utility/strings.h"
+#include "azure_c_shared_utility/xlogging.h"
+
+// DPS related header files
+#include "azure_prov_client/iothub_security_factory.h"
+#include "azure_prov_client/prov_device_client.h"
+#include "azure_prov_client/prov_transport_mqtt_client.h"
+#include "azure_prov_client/prov_security_factory.h"
+
+// Global provisioning endpoint for DPS
+static const char* g_dps_GlobalProvUri = "global.azure-devices-provisioning.net";
+// Format of custom DPS payload sent when registering a PnP device.
+static const char g_dps_PayloadFormatForModelId[] = "{\"modelId\":\"%s\"}";
+
+// State of DPS registration process.  We cannot proceed with PnP until we get into the state PNP_DPS_REGISTRATION_SUCCEEDED.
+typedef enum PNP_DPS_REGISTRATION_STATUS_TAG
+{
+    PNP_DPS_REGISTRATION_PENDING,
```

Comment:
should there be a `NOT_STARTED` or something similar? Because is it really `PENDING` if the provisioning request hasn't been fired off yet?

Reply:
You could argue it's pending because it's not done yet regardless of being started:).   

Adding a new state to this could end up with an additional level of indentation in `PnP_CreateDeviceClientHandle_ViaDps`.  I'm going to do instead is rename this to `PNP_DPS_REGISTRATION_NOT_COMPLETE`.  I'm also resetting `g_pnpDpsRegistrationStatus` to this initial state at top of `PnP_CreateDeviceClientHandle_ViaDps` just in case an app called it again.